<a href="https://colab.research.google.com/github/ManalHajj/Credit_Card_Fraud_Detection/blob/main/Credit_Card_Fraud_Detection.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

In [1]:
import numpy as np
import pandas as pd 
from sklearn.model_selection import train_test_split
from sklearn.linear_model import LogisticRegression
from sklearn.metrics import accuracy_score

In [18]:
# loading our dataset to a Pandas dataframe
credit_card_data = pd.read_csv('/content/creditcard.csv')

In [19]:
credit_card_data.head()

Time        V1        V2        V3        V4        V5        V6        V7  \
0     0 -1.359807 -0.072781  2.536347  1.378155 -0.338321  0.462388  0.239599   
1     0  1.191857  0.266151  0.166480  0.448154  0.060018 -0.082361 -0.078803   
2     1 -1.358354 -1.340163  1.773209  0.379780 -0.503198  1.800499  0.791461   
3     1 -0.966272 -0.185226  1.792993 -0.863291 -0.010309  1.247203  0.237609   
4     2 -1.158233  0.877737  1.548718  0.403034 -0.407193  0.095921  0.592941   

         V8        V9  ...       V21       V22       V23       V24       V25  \
0  0.098698  0.363787  ... -0.018307  0.277838 -0.110474  0.066928  0.128539   
1  0.085102 -0.255425  ... -0.225775 -0.638672  0.101288 -0.339846  0.167170   
2  0.247676 -1.514654  ...  0.247998  0.771679  0.909412 -0.689281 -0.327642   
3  0.377436 -1.387024  ... -0.108300  0.005274 -0.190321 -1.175575  0.647376   
4 -0.270533  0.817739  ... -0.009431  0.798278 -0.137458  0.141267 -0.206010   

        V26       V27       V28  Amount  Class  
0 -0.189115  0.133558 -0.021053  149.62    0.0  
1  0.125895 -0.008983  0.014724    2.69    0.0  
2 -0.139097 -0.055353 -0.059752  378.66    0.0  
3 -0.221929  0.062723  0.061458  123.50    0.0  
4  0.502292  0.219422  0.215153   69.99    0.0  

[5 rows x 31 columns]

Null values sum

In [20]:
credit_card_data.isnull().sum()

Time      0
V1        0
V2        0
V3        0
V4        0
V5        0
V6        0
V7        1
V8        1
V9        1
V10       1
V11       1
V12       1
V13       1
V14       1
V15       1
V16       1
V17       1
V18       1
V19       1
V20       1
V21       1
V22       1
V23       1
V24       1
V25       1
V26       1
V27       1
V28       1
Amount    1
Class     1
dtype: int64

In [21]:
credit_card_data.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 31780 entries, 0 to 31779
Data columns (total 31 columns):
 #   Column  Non-Null Count  Dtype  
---  ------  --------------  -----  
 0   Time    31780 non-null  int64  
 1   V1      31780 non-null  float64
 2   V2      31780 non-null  float64
 3   V3      31780 non-null  float64
 4   V4      31780 non-null  float64
 5   V5      31780 non-null  float64
 6   V6      31780 non-null  float64
 7   V7      31779 non-null  float64
 8   V8      31779 non-null  float64
 9   V9      31779 non-null  float64
 10  V10     31779 non-null  float64
 11  V11     31779 non-null  float64
 12  V12     31779 non-null  float64
 13  V13     31779 non-null  float64
 14  V14     31779 non-null  float64
 15  V15     31779 non-null  float64
 16  V16     31779 non-null  float64
 17  V17     31779 non-null  float64
 18  V18     31779 non-null  float64
 19  V19     31779 non-null  float64
 20  V20     31779 non-null  float64
 21  V21     31779 non-null  float64
 22

 Find the row with missing values

In [22]:
credit_card_data[credit_card_data.isnull().any(1)]


Time        V1        V2        V3        V4      V5       V6  V7  V8  \
31779  36513  1.295137  0.096775  0.348134  0.629896 -0.1727 -0.15435 NaN NaN   

       V9  ...  V21  V22  V23  V24  V25  V26  V27  V28  Amount  Class  
31779 NaN  ...  NaN  NaN  NaN  NaN  NaN  NaN  NaN  NaN     NaN    NaN  

[1 rows x 31 columns]

Delete the row with missing values

In [23]:
mod_df = credit_card_data.dropna()

In [24]:
mod_df.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 31779 entries, 0 to 31778
Data columns (total 31 columns):
 #   Column  Non-Null Count  Dtype  
---  ------  --------------  -----  
 0   Time    31779 non-null  int64  
 1   V1      31779 non-null  float64
 2   V2      31779 non-null  float64
 3   V3      31779 non-null  float64
 4   V4      31779 non-null  float64
 5   V5      31779 non-null  float64
 6   V6      31779 non-null  float64
 7   V7      31779 non-null  float64
 8   V8      31779 non-null  float64
 9   V9      31779 non-null  float64
 10  V10     31779 non-null  float64
 11  V11     31779 non-null  float64
 12  V12     31779 non-null  float64
 13  V13     31779 non-null  float64
 14  V14     31779 non-null  float64
 15  V15     31779 non-null  float64
 16  V16     31779 non-null  float64
 17  V17     31779 non-null  float64
 18  V18     31779 non-null  float64
 19  V19     31779 non-null  float64
 20  V20     31779 non-null  float64
 21  V21     31779 non-null  float64
 22

In [25]:
#number of legit and fraudulent transactions
mod_df['Class'].value_counts()

0.0    31677
1.0      102
Name: Class, dtype: int64

In [26]:
# seperating data for analysis
legit = mod_df[mod_df.Class == 0]
fraudulent = mod_df[mod_df.Class == 1]

In [27]:
print(legit.shape)
print(fraudulent.shape)

(31677, 31)
(102, 31)


In [28]:
legit.Amount.describe()

count    31677.000000
mean        81.082407
std        223.072655
min          0.000000
25%          6.870000
50%         20.000000
75%         73.610000
max       7879.420000
Name: Amount, dtype: float64

In [29]:
fraudulent.Amount.describe()

count     102.000000
mean       91.237451
std       248.270971
min         0.000000
25%         1.000000
50%         3.440000
75%        99.990000
max      1809.680000
Name: Amount, dtype: float64

In [30]:
# comparing
mod_df.groupby('Class').mean()

Time        V1        V2         V3        V4        V5  \
Class                                                                    
0.0    22335.407141 -0.188381  0.085855   0.759238  0.192821 -0.192581   
1.0    20334.872549 -7.761095  5.946853 -11.035230  5.926171 -5.529543   

             V6        V7        V8        V9  ...       V20      V21  \
Class                                          ...                      
0.0    0.095930 -0.095538  0.019413  0.324683  ...  0.041991 -0.03532   
1.0   -2.295415 -7.767215  3.904551 -2.997290  ...  0.679987  0.64190   

            V22       V23       V24       V25       V26       V27       V28  \
Class                                                                         
0.0   -0.118406 -0.039544  0.009212  0.133785  0.022140  0.010340  0.003013   
1.0   -0.352131 -0.317568 -0.241053  0.314412  0.177056  0.811372  0.118708   

          Amount  
Class             
0.0    81.082407  
1.0    91.237451  

[2 rows x 30 columns]

To deal with the high unbalance and make the ditribution even we're going to build a sample dataset containing each 102 fraudulent and legit transactions

In [112]:
legit_sample = legit.sample(n=102)

In [113]:
new_df = pd.concat([legit_sample,fraudulent],axis=0)
#axis 0 means rows

In [115]:
new_df.head()

Time        V1        V2        V3        V4        V5        V6  \
13847  24588 -2.878059  2.702684  0.370795 -1.836653 -0.291446 -0.063636   
25720  33734  1.386725 -0.606973 -0.165294 -0.941099 -0.524494 -0.355473   
24730  33355  0.543371 -0.872122  0.158282  1.577643 -0.609873  0.090597   
28811  35172  1.242540  1.038273 -2.120796  1.225245  1.486180 -0.412049   
23529  32803 -0.774519  0.948099  1.891155 -1.443315 -0.057959 -0.819604   

             V7        V8        V9  ...       V21       V22       V23  \
13847 -0.111826 -0.498550  3.073931  ...  0.278879 -0.738700  0.198426   
25720 -0.481989 -0.065100 -1.024421  ...  0.219095  0.446239 -0.253066   
24730  0.259075  0.123713  0.202201  ...  0.123321 -0.177543 -0.289267   
28811  0.446098  0.078068 -0.352031  ... -0.229004 -0.643843 -0.313630   
23529  0.792902 -0.184547 -0.114821  ... -0.088090  0.055429 -0.108748   

            V24       V25       V26       V27       V28  Amount  Class  
13847 -0.378549  0.205618  0.766655  1.238997  0.663801    0.77    0.0  
25720 -0.499811  0.688950 -0.082177 -0.019820 -0.004267   40.85    0.0  
24730  0.207547  0.449587 -0.368520 -0.028090  0.055426  299.75    0.0  
28811 -1.524499  0.848287 -0.203413  0.039089  0.067589    1.00    0.0  
23529  0.594801 -0.341317  0.661488  0.204506  0.004865    1.31    0.0  

[5 rows x 31 columns]

In [116]:
new_df['Class'].value_counts()

0.0    102
1.0    102
Name: Class, dtype: int64

In [114]:
new_df.groupby('Class').mean()

Time        V1        V2         V3        V4        V5  \
Class                                                                    
0.0    23926.637255 -0.144121  0.160171   0.658023  0.296444 -0.200729   
1.0    20334.872549 -7.761095  5.946853 -11.035230  5.926171 -5.529543   

             V6        V7        V8        V9  ...       V20       V21  \
Class                                          ...                       
0.0    0.069859 -0.064950  0.074292  0.303328  ...  0.086842 -0.011979   
1.0   -2.295415 -7.767215  3.904551 -2.997290  ...  0.679987  0.641900   

            V22       V23       V24       V25       V26       V27       V28  \
Class                                                                         
0.0   -0.155031 -0.040796 -0.021939  0.129880 -0.010175  0.054393  0.032632   
1.0   -0.352131 -0.317568 -0.241053  0.314412  0.177056  0.811372  0.118708   

          Amount  
Class             
0.0    82.633529  
1.0    91.237451  

[2 rows x 30 columns]

Splitting the data into features & targets 

In [117]:
X = new_df.drop(columns='Class',axis=1)
Y = new_df['Class']

In [118]:
print(X)

        Time        V1        V2        V3        V4        V5        V6  \
13847  24588 -2.878059  2.702684  0.370795 -1.836653 -0.291446 -0.063636   
25720  33734  1.386725 -0.606973 -0.165294 -0.941099 -0.524494 -0.355473   
24730  33355  0.543371 -0.872122  0.158282  1.577643 -0.609873  0.090597   
28811  35172  1.242540  1.038273 -2.120796  1.225245  1.486180 -0.412049   
23529  32803 -0.774519  0.948099  1.891155 -1.443315 -0.057959 -0.819604   
...      ...       ...       ...       ...       ...       ...       ...   
30398  35906 -3.519030  4.140867 -3.628202  5.505672 -4.057463 -0.905945   
30442  35926 -3.896583  4.518355 -4.454027  5.547453 -4.121459 -1.163407   
30473  35942 -4.194074  4.382897 -5.118363  4.455230 -4.812621 -1.224645   
30496  35953 -4.844372  5.649439 -6.730396  5.252842 -4.409566 -1.740767   
31002  36170 -5.685013  5.776516 -7.064977  5.902715 -4.715564 -1.755633   

             V7        V8        V9  ...       V20       V21       V22  \
13847 -0.1118

In [119]:
print(Y)

13847    0.0
25720    0.0
24730    0.0
28811    0.0
23529    0.0
        ... 
30398    1.0
30442    1.0
30473    1.0
30496    1.0
31002    1.0
Name: Class, Length: 204, dtype: float64


Splitting the data into Training & Testing data

In [120]:
X_train, X_test, Y_train, Y_test = train_test_split(X, Y, test_size=0.2, stratify=Y, random_state=2)

In [121]:
print(X.shape,X_train.shape,X_test.shape)

(204, 30) (163, 30) (41, 30)


we have 204 in our dataset 163 goes *to* training and the rest ( 4 ) to testing

Model Training 

Logistic Regression

In [122]:
model = LogisticRegression()

In [123]:
# Training the logistic regression model
model.fit(X_train,Y_train)

/usr/local/lib/python3.7/dist-packages/sklearn/linear_model/_logistic.py:818: ConvergenceWarning: lbfgs failed to converge (status=1):
STOP: TOTAL NO. of ITERATIONS REACHED LIMIT.

Increase the number of iterations (max_iter) or scale the data as shown in:
    https://scikit-learn.org/stable/modules/preprocessing.html
Please also refer to the documentation for alternative solver options:
    https://scikit-learn.org/stable/modules/linear_model.html#logistic-regression
  extra_warning_msg=_LOGISTIC_SOLVER_CONVERGENCE_MSG,


LogisticRegression()

Model Evaluation

In [125]:
#Accuracy on training data
X_train_prediction=model.predict(X_train)
#compare value predicted by our model ( X_train_prediction ) with the original one in Y_train
training_data_accuracy=accuracy_score(X_train_prediction,Y_train)

In [126]:
print('The accuracy is : ', training_data_accuracy)

The accuracy is :  0.9815950920245399


In [131]:
#accuracy on test data 
X_test_prediction = model.predict(X_test)
test_data_accuracy =accuracy_score(X_test_prediction,Y_test)

In [132]:
print('The accuracy is : ', test_data_accuracy )

The accuracy is :  0.9024390243902439
